src/gt4py/analysis/passes.py

```diff
@@ -456,57 +472,194 @@ def __init__(self):
     def defaults(self):
         return self._DEFAULT_OPTIONS
 
-    def apply(self, transform_data: TransformData):
-        # Greedy strategy to merge multi-stages
-        merged_blocks = [transform_data.blocks[0]]
-        for candidate in transform_data.blocks[1:]:
-            merged = merged_blocks[-1]
-            if self._are_compatible_multi_stages(
-                merged, candidate, transform_data.has_sequential_axis
-            ):
-                merged.id = transform_data.id_generator.new
-                self._merge_domain_blocks(merged, candidate)
+    def _greedy_merging(
+        self,
+        items: Sequence[T],
+        can_merge: Callable[[T, T], bool],
+        do_merge: Callable[[T, T], None],
+    ) -> Sequence[T]:
+        merged_items = [items[0]]
+        for item in items[1:]:
+            last_merged = merged_items[-1]
+            if can_merge(last_merged, item):
+                do_merge(last_merged, item)
             else:
-                merged_blocks.append(candidate)
+                merged_items.append(item)
+        return merged_items
+
+    def _can_merge_multi_stages_callback(
+        self, transform_data: TransformData
+    ) -> Callable[[DomainBlockInfo, DomainBlockInfo], bool]:
+        def can_merge(last_merged: DomainBlockInfo, candidate: DomainBlockInfo) -> bool:
+            return self._are_compatible_multi_stages(
```

Comment:
None of the functions called in these callbacks have state, i.e. depend on self. Why not just encapsulate everything that follows below in this file into two classes one for stages and one for multistages, with a simple interface `can_merge` and `do_merge`. Then the two would be clearly separated and the pass itself would only be responsible for the merging strategy not actual machinery.